grpc/src/main/java/io/stargate/grpc/service/MessageStreamObserver.java

```diff
@@ -0,0 +1,133 @@
+ */
+public class MessageStreamObserver<MessageT extends GeneratedMessageV3>
+    implements StreamObserver<MessageT> {
+
+  static final int DELAY_BETWEEN_ON_COMPLETE_CHECKS_MS = 100;
+  private static final int ON_COMPLETE_RETRIES_DEFAULT =
+      Integer.getInteger("stargate.grpc.streaming.on_complete.retries", 100);
+
+  private final AtomicLong inFlight = new AtomicLong(0);
+  private final ExceptionHandler exceptionHandler;
+  private final BiFunction<MessageT, AtomicLong, MessageHandler<MessageT, ?>> handlerProducer;
+  private final StreamObserver<QueryOuterClass.Response> responseObserver;
+  private final ScheduledExecutorService executor;
+  private final int onCompleteRetries;
+
+  MessageStreamObserver(
+      StreamObserver<QueryOuterClass.Response> responseObserver,
+      ExceptionHandler exceptionHandler,
+      BiFunction<MessageT, AtomicLong, MessageHandler<MessageT, ?>> handlerProducer,
+      ScheduledExecutorService executor,
+      int onCompleteRetries) {
+    this.responseObserver = responseObserver;
+    this.exceptionHandler = exceptionHandler;
+    this.handlerProducer = handlerProducer;
+    this.executor = executor;
+    this.onCompleteRetries = onCompleteRetries;
+  }
+
+  public MessageStreamObserver(
+      StreamObserver<QueryOuterClass.Response> responseObserver,
+      ExceptionHandler exceptionHandler,
+      BiFunction<MessageT, AtomicLong, MessageHandler<MessageT, ?>> handlerProducer,
+      ScheduledExecutorService executor) {
+    this(
+        responseObserver, exceptionHandler, handlerProducer, executor, ON_COMPLETE_RETRIES_DEFAULT);
+  }
+
+  /**
+   * It is creating new handler for each request and invokes the {@link MessageHandler#handle()}
+   * method.
+   *
+   * @param value Batch or Query
+   */
+  @Override
+  public void onNext(MessageT value) {
+    inFlight.incrementAndGet();
+    handlerProducer.apply(value, inFlight).handle();
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    exceptionHandler.handleException(t);
+  }
+
+  /**
+   * It completes the response Observer. If there are any in-flight Queries, it waits until all of
+   * them are processed.
+   *
+   * <p>According to reactive contract:
+   *
+   * <p>An Observable may make zero or more OnNext notifications, each representing a single emitted
+   * item, and it may then follow those emission notifications by either an OnCompleted or an
+   * OnError notification, but not both.
+   *
+   * <p>If there was any error (server-side or client-side) we are ignoring the onComplete() signal.
+   */
+  @Override
+  public void onCompleted() {
+    CompletableFuture<Void> shouldComplete = new CompletableFuture<>();
+    if (exceptionHandler.getExceptionOccurred()) {
+      return;
+    }
+
+    executor.schedule(
+        () -> waitForAllInFlightProcessed(onCompleteRetries - 1, shouldComplete),
+        DELAY_BETWEEN_ON_COMPLETE_CHECKS_MS,
+        TimeUnit.MILLISECONDS);
+    try {
+      shouldComplete.get();
```

Comment:
What thread gets blocked here actually? And why do we actually block here..

Shouldn't this just be a signal from the client side there are no more requests? Meaning, hey there are no more requests coming, you can signal `onComplete` once you have processed all the in flight requests...

So if I may I would propose a design change here.. I would make `StreamingBatchHandler` and `StreamingQueryHandler` only pass the response back to the `MessageStreamObserver` and the stream observer can hold the whole state and act accordingly. This would also make `StreamingBatchHandler` and `StreamingQueryHandler` completely stateless, and would not have to do any `inFlight` stuff..

And I am certain this approach is actually a bug, why would you block the client side, I don't think this is going in line with the observable signaling..

Reply:
>Shouldn't this just be a signal from the client side there are no more requests? Meaning, hey there are no more requests coming, you can signal onComplete once you have processed all the in-flight requests

That is actually the case, the client is sending `onComplete` once all `onNext()` are emitted:
see for example: `simpleReactiveBiDirectionalQueries()`:
```
   emitter.next(
                  cqlQuery("INSERT INTO test (k, v) VALUES ('a', 1)", queryParameters(keyspace)));
              emitter.next(
                  cqlQuery(
                      "INSERT INTO test (k, v) VALUES (?, ?)",
                      queryParameters(keyspace),
                      Values.of("b"),
                      Values.of(2)));
              emitter.complete(); // client-side complete signal
```
regarding blocking `onComplete()`:

I first implemented the `onComplete` as not-blocking and without any in-flight.
However, all `next()` and once the `onComplete()` is issued by the caller, there still might be not-processed requests.
Without waiting for in-flight, that not-processed ' onNext()`s are not propagated to the caller because the stream will be closed by `complete()`. This is why decided to wait in `onComplete()`
(see this for more info: https://github.com/stargate/stargate/pull/1469#discussion_r772934106)
The other solution would be to impose on the client to track all `onNext()` and only if all of them succeeded to call the `onComplete`. I am not sure if this has a good UX. Wdyt?

Reply:
But this is not what I am talking about.. You did good with the `inFlight` counting imo.. I just say let's switch the approach a bit:
* instead of blocking get here, set an `AtomicBoolean noMoreRequests` to true.. that's actually the only thing we need..
* change the `Streaming..Handler`s to delegate back the response to the `MessageStreamObserver` in the `setSuccess`
* based on the current state of the `noMoreRequests` and `inFlight` make a decision what to do

Reply:
done in: 
https://github.com/stargate/stargate/pull/1469/commits/825c3dd263fa3b6798290ad059126fd32f059e16